src/main/java/com/google/codeu/servlets/MessageServlet.java

```diff
@@ -118,5 +99,40 @@ public boolean isValidURL(String url) {
       return false;
     }
   }
+
+  public String displayMedia(String regexURL, String userInput) {
+    String replacement;
+    String textWithMediaReplaced = userInput;
+
+    Pattern pattern = Pattern.compile(regexURL);
+    Matcher matcher = pattern.matcher(userInput);
+
+    while (matcher.find()) {
+      String mediaURL = matcher.group(1);
+      System.out.println("A media url found: " + mediaURL);
+
+      if (isValidURL(mediaURL)) {
+        if (mediaURL.endsWith(".mp3")) {
+          replacement = "<audio controls src=" + mediaURL + " />";
+          textWithMediaReplaced = textWithMediaReplaced.replace(mediaURL, replacement);
+          System.out.println("URL changed with audio tag: " + textWithMediaReplaced);
+        } else if (mediaURL.endsWith(".mp4")) {
+          replacement = "<video controls src=" + mediaURL + " />";
+          textWithMediaReplaced = textWithMediaReplaced.replace(mediaURL, replacement);
+          System.out.println("URL changed with video tag: " + textWithMediaReplaced);
+        } else {
+          replacement = "<img src=" + mediaURL + " />";
+          textWithMediaReplaced = textWithMediaReplaced.replace(mediaURL, replacement);
+          System.out.println("URL changed with image tag: " + textWithMediaReplaced);
+        }
+      } else {
+        replacement = mediaURL + " (Not a valid URL)";
+        textWithMediaReplaced = textWithMediaReplaced.replace(mediaURL, replacement);
```

Comment:
if it's not valid, should we just keep the input as-is?

Reply:
Also can! I was thinking the user could use a note to know why it wasn't rendering haha but I think as we improve it and depending on our decided mode of image can consider nicer looking ways to display as well

Reply:
Ah that's reasonable too, this is fine.